chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+The recipe file entered by the user declares all the steps taken in the
+simulation and the dependencies in form of exchanged data
+between the steps. The Recipe module holds all classes and functions
+needed to parse a json file into a recipe object and check integrity.
+"""
+# classes to put recipe json data into
+
+
+from __future__ import unicode_literals
+import os
+import io
+import platform
+import json
+import sys
+
+# logs need to be imported this way to not write logs.logger all the time
+from .logs import *
+
+# constants
+# built-in functions that can be called as a simulation step inside a node
+BUILT_INS = ["collect"]
+
+
+class Recipe:
+    """
+    A recipe is the workflow representation of a simulation. It is struktured
+    as a list of nodes of class Node. The nodes already contain information
+    about the data flow and dependencies in the workflow, so there is not
+    explicit representation of edges or dependencies needed.
+    """
+    nodes = []
+
+    def __init__(self, nodelist):
+        """
+        Initialises a recipe by appending the `nodelist` to `nodes`.
+
+        :param nodelist: list of simulation steps (nodes)
+        :type nodelist: Node[]
+        """
+        self.nodes = nodelist
+        # Making sure, that nodelist is a list of type Node
+        try:
+            emptyNode = Node(
+                "nodeName",
+                {"i": "input"},
+                {"o": "output"},
+                "collect")
+            self.nodes.append(emptyNode)
+            self.nodes.pop()
+        except Exception as exc:         # mit Klasse
+            logger.exception(exc)
+            #raise TypeError("""Class Recipe expects a list to be
+            #    initialised with.""")
+
+    def inputIntegrity(self):
+        """
+        Tests if there is exactly one incoming edge to every input.
+        Warns, if there is no incoming edge. Excludes nodes from recipe,
+        that have no incoming edge for a node or have uncomputable inputs
+        because of missing inputs in parent nodes.
+        
+        :returns:
+            err - String. None if correct. Else error message. \n
+            warn - String. None if correct. Else error message.
+        """
+        err = None
+        warn = None
+        # 1. make unique list of outputs
+        outputs_of_all_nodes = []
+        for node in self.nodes:
+            for key in node.outputs:
+                output = node.outputs[key]
+                if output in outputs_of_all_nodes:
+                    # overwrite exception class to generate warns
+                    err = (("" if err is None else err) + 'The output ' +
+                           output + ' of node ' + node.name +
+                           ' has the same name as an output declared before. ')
+                else:
+                    outputs_of_all_nodes.append(output)
```

Comment:
Coding list processing like this unfortunately turns slow and difficult to read in python.
You might replace the above code with the suggestion below, that exploits some functionality of `set`s:
```
outputs_of_all_nodes = set([])
for node in self.nodes:
    keys = set(node.outputs)
    if len(keys & outputs_of_all_nodes) > 0:
        raise ValueError("Multiple outputs blah")
    
    outputs_of_all_nodes.add(keys)
```

Reply:
The problem is not the use of the list itself but the fact, that I need to compute a union of the values inside `self.nodes` and `outputs_of_all_nodes`. Until I find a way to invert a dictionary, it won't save a lot of lines.